usd/src/UsdError.cc

```diff
@@ -0,0 +1,176 @@
+    : dataPtr(ignition::utils::MakeImpl<Implementation>())
+{
+  this->dataPtr->code = _code;
+  this->dataPtr->message = _message;
+}
+
+/////////////////////////////////////////////////
+UsdError::UsdError(const UsdErrorCode _code, const std::string &_message,
+                   const std::string &_filePath)
+    : dataPtr(ignition::utils::MakeImpl<Implementation>())
+{
+  this->dataPtr->code = _code;
+  this->dataPtr->message = _message;
+  this->dataPtr->filePath = _filePath;
+}
+
+/////////////////////////////////////////////////
+UsdError::UsdError(const UsdErrorCode _code, const std::string &_message,
+                   const std::string &_filePath, int _lineNumber)
+    : dataPtr(ignition::utils::MakeImpl<Implementation>())
+{
+  this->dataPtr->code = _code;
+  this->dataPtr->message = _message;
+  this->dataPtr->filePath = _filePath;
+  this->dataPtr->lineNumber = _lineNumber;
+}
+
+/////////////////////////////////////////////////
+UsdError::UsdError(const sdf::Error &_sdf_error)
+    : dataPtr(ignition::utils::MakeImpl<Implementation>())
+{
+  this->dataPtr->code = UsdErrorCode::SDF_ERROR;
+  this->dataPtr->sdf_error = _sdf_error;
+}
+
+/////////////////////////////////////////////////
+UsdErrorCode UsdError::Code() const
+{
+  return this->dataPtr->code;
+}
+
+/////////////////////////////////////////////////
+std::string UsdError::Message() const
+{
+  return this->dataPtr->message;
+}
+
+/////////////////////////////////////////////////
+std::optional<std::string> UsdError::FilePath() const
+{
+  return this->dataPtr->filePath;
+}
+
+/////////////////////////////////////////////////
+void UsdError::SetFilePath(const std::string &_filePath)
+{
+  this->dataPtr->filePath = _filePath;
+}
+
+/////////////////////////////////////////////////
+std::optional<int> UsdError::LineNumber() const
+{
+  return this->dataPtr->lineNumber;
+}
+
+/////////////////////////////////////////////////
+void UsdError::SetLineNumber(int _lineNumber)
+{
+  this->dataPtr->lineNumber = _lineNumber;
+}
+
+/////////////////////////////////////////////////
+std::optional<sdf::Error> UsdError::SdfError() const
+{
+  return this->dataPtr->sdf_error;
+}
+
+/////////////////////////////////////////////////
+UsdError::operator bool() const
+{
+  return this->dataPtr->code != UsdErrorCode::NONE;
+}
+
+/////////////////////////////////////////////////
+bool UsdError::operator==(const bool _value) const
+{
+  return ((this->dataPtr->code != UsdErrorCode::NONE) && _value) ||
+         ((this->dataPtr->code == UsdErrorCode::NONE) && !_value);
+}
+
+namespace sdf
+{
+// Inline bracket to help doxygen filtering.
+inline namespace SDF_VERSION_NAMESPACE
+{
+namespace usd
+{
+
+/////////////////////////////////////////////////
+std::ostream &operator<<(std::ostream &_out, const sdf::usd::UsdError &_err)
+{
+  if (_err.dataPtr->code == UsdErrorCode::SDF_ERROR)
+  {
+    _out << _err.dataPtr->sdf_error.value();
```

Comment:
There's a code path through the constructor where someone might accidentally set the code to `SDF_ERROR` without assigning `sdf_error`, so consider checking if `sdf_error` has a value before using it?

Reply:
Nice catch! I handled this in https://github.com/ignitionrobotics/sdformat/pull/836/commits/287fa31e0ddbdfcecd3309e108b459b00c21515f. I also documented the constructors to note that if the error code is `SDF_ERROR`, the constructor that takes an `sdf::Error` object should be used.

Reply:
LGTM!